text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,411 @@
+                Field::new("span_id", DataType::Binary, false),
+                Field::new("trace_state", DataType::Utf8, false),
+                Field::new("attributes", DataType::List(
+                    Box::new(Field::new(
+                        "attribute",
+                        DataType::Struct(vec![
+                            Field::new("name", DataType::Utf8, false),
+                            Field::new("value", DataType::Utf8, false),
+                        ]),
+                        true,
+                    ))
+                ), true),
+            ]),
+            true,
+        ))
+    ), true),
+])
+```
+
+### Corners cases
+
+Backends that don't support natively multivariate time-series can still automatically transform these events in multiple univariate time-series and operate as usual.
+
+Specialized processors can be developed to group metrics, logs and traces in optimized batch of events in order to connect existing OTEL collectors with backends supporting this new protocol extension.
+
+Specialized processors can be developed to convert batch of events into the existing entities for backends that don't support this protocol extension.
+
+## Trade-offs and mitigations
+
+A columnar-oriented protocol is not necessarily desirable for all scenarios (e.g. devices that do not have the resources to accumulate data in batches). The proposed mixed solution allows to better address these different scenarios.
+
+Implementing a new columnar format is complex and costly (multiple language implementations, tests, optimizations, industry adoption). Reusing Apache Arrow is an interesting approach to mitigate this issue.
+
+## Prior art and alternatives
+
+* [Column-oriented DBMS](https://en.wikipedia.org/wiki/Column-oriented_DBMS)
+* [Apache Arrow](https://arrow.apache.org/)
+
+## Open questions
+
+More work needs to be done around examplars and histograms representation.
+
+More discussions should happen on the processor and storage layers. This approach could simplify significantly the design of OTEL compatible stream processing and database systems.
+
+## Future possibilities
+
+* Leverage Apache Arrow dictionary
+
+## Appendices
+
+### event proto
+
+Protobuf specification (draft) for an Arrow-based OpenTelemetry event.
+
+```protobuf
+syntax = "proto3";
+
+package opentelemetry.proto.arrow_events.v1;
+
+import "opentelemetry/proto/common/v1/common.proto";
+import "opentelemetry/proto/resource/v1/resource.proto";
+
+option java_multiple_files = true;
+option java_package = "io.opentelemetry.proto.events.v1";
+option java_outer_classname = "EventsProto";
+option go_package = "github.com/open-telemetry/opentelemetry-proto/gen/go/events/v1";
+
+// A column-oriented collection of events from a Resource.
+message ResourceEvents {
+  // The resource for the events in this message.
+  // If this field is not set then no resource info is known.
+  opentelemetry.proto.resource.v1.Resource resource = 1;
```

Comment:
Another thought experiment: why shouldn't the protocol use one giant arrow buffer to represent more than one library of data? (Answer: they have different schemas?)

Reply:
Yes they have different schemas.